azure-keyvault/azure/keyvault/custom/key_vault_authentication.py

```diff
@@ -19,27 +16,92 @@ class KeyVaultAuthBase(AuthBase):
     def __init__(self, authorization_callback):
         self._callback = authorization_callback
         self._token = None
+        self._thread_local = threading.local()
+        self._thread_local.pos = None
+        self._thread_local.auth_attempted = False
 
     def __call__(self, request):
         # attempt to pre-fetch challenge if cached
         if self._callback:
             challenge = ChallengeCache.get_challenge_for_url(request.url)
             if challenge:
-                # if challenge cached, retrieve token and update the request
+                # if challenge cached, use the authorization_callback to retrieve token and update the request
                 self.set_authorization_header(request, challenge)
             else:
-                # send the request to retrieve the challenge, then request the token and update
-                # the request
-                # TODO: wire up commons flag for things like Fiddler, logging, etc.
-                response = requests.Session().send(request)
-                if response.status_code == 401:
-                    auth_header = response.headers['WWW-Authenticate']
-                    if HttpBearerChallenge.is_bearer_challenge(auth_header):
-                        challenge = HttpBearerChallenge(response.request.url, auth_header)
-                        ChallengeCache.set_challenge_for_url(response.request.url, challenge)
-                        self.set_authorization_header(request, challenge)
+                # if the challenge is not cached we will let the request proceed without the auth header so we
+                # get back the proper challenge in response. We register a callback to handle the response 401 response.
+                try:
+                    self._thread_local.pos = request.body.tell()
+                except AttributeError:
+                    self._thread_local.pos = None
+
+                self._thread_local.auth_attempted = False
+                request.register_hook('response', self.handle_401)
+                request.register_hook('response', self.handle_redirect)
+
         return request
 
+    def handle_redirect(self, r, **kwargs):
+        """Reset auth_attempted on redirects."""
+        if r.is_redirect:
+            self._thread_local.auth_attempted = False
+
+    def handle_401(self, response, **kwargs):
+        """
+        Takes the response authenticates and resends if neccissary
+        
+        :return: The final response to the authenticated request
+        :rtype: requests.Response
+        """
+        # If response is not 4xx do not auth and return response
```

Comment:
In [RFC 7235](https://tools.ietf.org/html/rfc7235#section-4.1)  it says:  A server MAY generate a WWW-Authenticate header field in other response messages to indicate that supplying credentials (or different credentials) might affect the response.  So I was going by that and what DigestAuth is doing in the request library https://github.com/requests/requests/blob/master/requests/auth.py#L231.  But if the data plane service will only challenge with a 401 response code we can restrict this to 401.